<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="GL-2.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Take Home<br>(Session 6 & 7)</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [1]:
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import statsmodels
import statsmodels.api as sm

# import 'stats' package from scipy library
from scipy import stats

# import statistics to perform statistical computations
import statistics

# to test the normality 
from scipy.stats import shapiro

# import a library to perform Z-test
from statsmodels.stats import weightstats as stests

# import statsmodels
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# import the functions to perform Chi-square tests
from scipy.stats import chi2_contingency
from scipy.stats import chisquare

# function to perform post-hoc test
import statsmodels.stats.multicomp as mc

In [2]:
# set the plot size using 'rcParams'
# once the plot size is set using 'rcParams', it sets the size of all the forthcoming plots in the file
# pass width and height in inches to 'figure.figsize' 
plt.rcParams['figure.figsize'] = [15,8]

### Let's begin with some hands-on practice exercises

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>1. The sales manager at an electronics company wants to study if online sales of their products in 2019 follow the same distribution as the online sales of products in 2015. He collected the sample of 300 products sold in 2019. Help the manager to test his claim using p-value technique at a 99% confidence level.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

<img src='chi_2.png'>

The null and alternative hypothesis is:

H<sub>0</sub>: There is no significant difference between sales in 2015 and 2019. <br>
H<sub>1</sub>: There is a significant difference between sales in 2015 and 2019.

In [21]:
# given observed values
observed_value = [85, 90, 125]

# expected % sale in 2015 
exp_count = [0.38, 0.20, 0.42]

# calculate the expected values for each category
expected_value = np.array(exp_count) * 300

# use the 'chisquare()' to perform the goodness of fit test
# the function returns the test statistic value and corresponding p-value
# pass the observed values to the parameter, 'f_obs'
# pass the expected values to the parameter, 'f_exp'
stat, p_value = chisquare(f_obs = observed_value, f_exp = expected_value)

# print the p-value
print('p-value:', p_value)

p-value: 1.3776246111575923e-05


The above output shows that the p-value is less than 0.01. Thus, we reject the null hypothesis and conclude that there is a significant difference between sales in 2015 and 2019.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>2. Use Chi Square testing to check if the category  of car maintenance is dependent on the type of car's transmission. Test the hypothesis with 95% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

<img src='chi_ind.png'>

The null and alternative hypothesis is:

H<sub>0</sub>: Type of car maintenance and type of transmission is independent<br>
H<sub>1</sub>: Type of car maintenance and type of transmission is not independent

There are 3 types of cars and two types of transmissions. Thus degrees of freedom are (3-1)(2-1) = 2. And given level of significance is 0.05.

In [22]:
# observed values  
observed_value = np.array([[104, 85, 78], [35, 61, 72]])
observed_value

array([[104,  85,  78],
       [ 35,  61,  72]])

In [23]:
# use the 'chi2_contingency()' to check the independence of variables
# the function returns the test statistic value, corresponding p-value, degrees of freedom of the test and expected values
# pass the observed values to the parameter, 'observed'
# 'correction = False' will not apply the Yates' correction
test_stat, p, dof, expected_value = chi2_contingency(observed = observed_value, correction = False)

print("Test statistic:", round(test_stat,3))
print("p-value:",round(p,4))
print("Degrees of freedom:", dof)
print("Expected values:", expected_value)

Test statistic: 16.775
p-value: 0.0002
Degrees of freedom: 2
Expected values: [[85.31724138 89.6137931  92.06896552]
 [53.68275862 56.3862069  57.93103448]]


The above output shows that the p-value is less than 0.05. Thus, we reject the null hypothesis and conclude that there is a significant difference between type of car maintenance and transmission.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>3. The company claims that in the organization, 68% of employees are there with rare travelling, 20% with frequent travelling and 12% of employees does not need to travel for business. Check if the given data fits with the company's claimed distribution. Use the p-value technique with 99% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data available in the CSV file `Employee_Attrition.csv`.

In [25]:
# read the sample data 
df_emp = pd.read_csv('Employee_Attrition.csv')

# display the first two observations
df_emp.head(2)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  

[2 rows x 35 columns]

The null and alternative hypothesis is:

H<sub>0</sub>: There is no significant difference between the observed and expected values. <br>
H<sub>1</sub>: There is a significant difference between the observed and expected values.

In [26]:
# observed values
df_emp['BusinessTravel'].value_counts()

Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64

In [27]:
# given observed values
observed_value = [1043, 277, 150]

# expected percentage 
exp_count = [0.68, 0.20, 0.12]

# calculate the expected values for each category
expected_value = np.array(exp_count) * len(df_emp)

# round-off the values to integers
expected_value = round(pd.Series(expected_value))

# use the 'chisquare()' to perform the goodness of fit test
# the function returns the test statistic value and corresponding p-value
# pass the observed values to the parameter, 'f_obs'
# pass the expected values to the parameter, 'f_exp'
stat, p_value = chisquare(f_obs = observed_value, f_exp = expected_value)


# use round function to round off the values
p_value = round(p_value,3)


print('p-value:', p_value)

p-value: 0.036


The above output shows that the p-value is greater than 0.01. Thus, we fail to reject (i.e. accept) the null hypothesis and conclude that company's claim is correct. i.e. there is no significant difference between the observed and expected values.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                        <b>4. Check whether travelling for the business is related to the gender of an employee. Use the p-value technique to test the claim with 90% confidence. </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data available in the CSV file `Employee_Attrition.csv`.

The null and alternative hypothesis is:

H<sub>0</sub>: Business travel and gender of an employee are independent<br>
H<sub>1</sub>: Business travel and gender of an employee are not independent.

In [28]:
# use 'crosstab()' to create a table for each type of business travel and gender 
table = pd.crosstab(df_emp['BusinessTravel'], df_emp['Gender'])

# observed values  
observed_value = table.values
observed_value

array([[ 49, 101],
       [117, 160],
       [422, 621]], dtype=int64)

In [29]:
# use the 'chi2_contingency()' to check the independence of variables
# the function returns the test statistic value, corresponding p-value, degrees of freedom of the test and expected values
# pass the observed values to the parameter, 'observed'
# 'correction = False' will not apply the Yates' correction
test_stat, p, dof, expected_value = chi2_contingency(observed = observed_value, correction = False)

# use round function to round off the values
p = round(p,3)

print("p-value:", p)

p-value: 0.133


The above output shows that the p-value is greater than 0.1, thus we fail to reject (i.e. accept) the null hypothesis and conclude that the business travel is independent of the gender of an employee.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>5. A company produces leather belts on three different production lines. The number of defective belts produced by each line is recorded. Test if all the production lines produce on an average equal number of defective belts. Use the p-value technique to test at 5% level of significance.  </b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given in the excel file `defective_prod.xlsx`

In [30]:
# read the data 
df_prod = pd.read_excel('defective_prod.xlsx')

# display the first five observations
df_prod.head()

production_line  def_belts
0               A        5.0
1               C        2.0
2               B        3.0
3               A        4.0
4               C        1.0

The null and alternative hypothesis is:

H<sub>0</sub>: The average of defective belts for all the production lines is same<br>
H<sub>1</sub>: At least one production line has a different average of defective belts

Let us check the normality of the number of defective belts from all the production lines.

In [31]:
# perform Shapiro-Wilk test to test the normality
# shapiro() returns a tuple having the values of test statistics and the corresponding p-value
stat, p_value = stats.shapiro(df_prod['def_belts'])

# round funtion to round off 3 digits
p_value = round(p_value,3)

# print the p-value
print('p-value:', p_value)

p-value: 0.114


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the number of defective belts is normally distributed. Thus the assumption of normality is satisfied.

Let us check the equality of variances.

In [32]:
# perform Levene's test for the equality of variances 
# levene() returns a tuple having the values of test statistics and the corresponding p-value
# pass the number of defective belts from each group
stat, p_value = stats.levene(df_prod[df_prod['production_line'] == 'A']['def_belts'], 
                             df_prod[df_prod['production_line'] == 'B']['def_belts'],
                             df_prod[df_prod['production_line'] == 'C']['def_belts'])

# round funtion to round off 3 digits
p_value = round(p_value,3)

# print the p-value 
print('P-Value:', p_value)

P-Value: 0.449


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the population variances are equal for all the samples.

Let us perform one-way ANOVA.

In [33]:
# perform one-way ANOVA
# pass the given data
f_test, p_val = stats.f_oneway(df_prod[df_prod['production_line'] == 'A']['def_belts'], 
                               df_prod[df_prod['production_line'] == 'B']['def_belts'],
                               df_prod[df_prod['production_line'] == 'C']['def_belts'])
# Use round funtion to round off the values
f_test = round(f_test,3)
p_val = round(p_val,3)
# print the test statistic and corresponding p-value
print('Test statistic:', f_test)
print('p-value:', p_val)

Test statistic: 0.838
p-value: 0.449


The above output shows that, the p-value is greater than 0.05. Thus we fail to reject (i.e. accept) the null hypothesis and conclude that the average of defective belts produced on all the production lines is the same.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>6. Check whether we can use the given dataset to study the equality of average monthly income of sales executives with a different education background in the company. Use a p-value technique to test at a 1% level of significance.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data available in the CSV file `sales_emp.csv`.

In [34]:
# read the sample data 
df_emp = pd.read_csv('sales_emp.csv')

# display the first two observations
df_emp.head(2)

Age BusinessTravel  DailyRate  DistanceFromHome  Education  \
0   36  Travel_Rarely       1218                 9          4   
1   39  Travel_Rarely        895                 5          3   

     EducationField  EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  \
0     Life Sciences              1              27                        3   
1  Technical Degree              1              42                        4   

   HourlyRate  ...  RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0          82  ...                         2             80                 0   
1          56  ...                         3             80                 1   

  TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                10                      4                3               5   
1                19                      6                4               1   

  YearsInCurrentRole YearsSinceLastPromotion  YearsWithCurrManager  
0                  3                       0                     3  
1                  0                       0                     0  

[2 rows x 31 columns]

In the given dataset, the `EducationField` is a categorical variable and `MonthlyIncome` is a numeric variable. To use the one-way ANOVA test for the equality of average income for all the education fields; the sample data should be drawn from normally distributed population. Also the population variances for all the groups should be equal.

Let us check the normality of the monthly income.

In [35]:
# perform Shapiro-Wilk test to test the normality
# shapiro() returns a tuple having the values of test statistics and the corresponding p-value
stat, p_value = stats.shapiro(df_emp['MonthlyIncome'])

# use round function to round off the values
p_value = round(p_value,3)

# print the p-value
print('p-value:', p_value)

p-value: 0.057


From the above result, we can see that the p-value is greater than 0.01, thus we can say that the monthly income is normally distributed. Thus the assumption of normality is satisfied.

Let us check the equality of variances.

In [36]:
# education fields in the data
df_emp['EducationField'].unique()

array(['Life Sciences', 'Technical Degree', 'Marketing', 'Medical',
       'Other'], dtype=object)

In [37]:
# perform Levene's test for the equality of variances 
# levene() returns a tuple having the values of test statistics and the corresponding p-value
# pass the number of defective belts from each group
stat, p_value = stats.levene(df_emp[df_emp['EducationField'] == 'Life Sciences']['MonthlyIncome'],
                             df_emp[df_emp['EducationField'] == 'Technical Degree']['MonthlyIncome'],
                             df_emp[df_emp['EducationField'] == 'Marketing']['MonthlyIncome'],
                             df_emp[df_emp['EducationField'] == 'Medical']['MonthlyIncome'],
                             df_emp[df_emp['EducationField'] == 'Other']['MonthlyIncome'])

# use round function to round off the values
p_value = round(p_value,3)

# print the p-value 
print('P-Value:', p_value)

P-Value: 0.235


From the above result, we can see that the p-value is greater than 0.01, thus we can say that the population variances are equal for all the groups.

Thus we can use the one-way ANOVA test to check the equality of average monthly income for all the education fields.

 <table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>7. Use the sales employees' dataset to test whether the average monthly income of sales executives with different education background is equal or not. Use a critical value method with 99% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

The null and alternative hypothesis is:

H<sub>0</sub>: The average monthly income for all the education fields is the same<br>
H<sub>1</sub>: The average monthly income due to at least one education field is different.

In [38]:
# number of education fields in the data
df_emp['EducationField'].nunique()

5

In [39]:
# total sample observations
len(df_emp)

54

Here t (=number of education fields) = 5, N (=total observations) = 54 

For ⍺ = 0.05 and degrees of freedom = (t-1, N-t) = (4, 49), calculate the critical value.

In [40]:
# calculate the F-value for 95% of confidence level
# use 'stats.f.isf()' to find the F-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.01
# pass the degrees of freedom (= 4) to the parameter 'dfn' 
# pass the degrees of freedom (= 49) to the parameter 'dfd' 
# use 'round()' to round-off the value to 4 digits
f = np.abs(round(stats.f.isf(q = 0.01, dfn = 4, dfd = 49), 4))

print('Critical value for F-test:', f)

Critical value for F-test: 3.7283


i.e. if the test statistic value is greater than 2.5611 then we reject the null hypothesis.

In [41]:
# perform one-way ANOVA
# pass the given data
test_stat, p_val = stats.f_oneway(df_emp[df_emp['EducationField'] == 'Life Sciences']['MonthlyIncome'],
                                  df_emp[df_emp['EducationField'] == 'Technical Degree']['MonthlyIncome'],
                                  df_emp[df_emp['EducationField'] == 'Marketing']['MonthlyIncome'],
                                  df_emp[df_emp['EducationField'] == 'Medical']['MonthlyIncome'],
                                  df_emp[df_emp['EducationField'] == 'Other']['MonthlyIncome'])

# use round function to round off the values
test_stat = round(test_stat,3)

# print the test statistic
print('Test statistic:', test_stat)

Test statistic: 0.955


The above output shows that the test statistic is less than 3.7283. Thus we fail to reject (i.e. accept) the null hypothesis and conclude that the average monthly income for all the education fields is the same.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>8. The company wants to decide their advertisement strategies for their new smartwatch. The manager collects the advertising cost (in dollars) for four different mediums. Test if the company is spending on average an equal amount on all the advertising mediums using a critical value approach with 99% confidence.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

Use the data given in the excel file `ad_cost.xlsx`

In [42]:
# read the data 
df_cost = pd.read_excel('ad_cost.xlsx')

# display the first five observations
df_cost.head()

medium   cost
0  Snapmart  252.0
1   Picbook  378.0
2    Funsta  312.0
3   Klicker  297.0
4   Picbook  361.0

The null and alternative hypothesis is:

H<sub>0</sub>: The average count for all the advertising mediums is same<br>
H<sub>1</sub>: At least one advertising medium has a different average cost

Let us check the normality of the cost for all the advertising mediums.

In [43]:
# perform Shapiro-Wilk test to test the normality
# shapiro() returns a tuple having the values of test statistics and the corresponding p-value
stat, p_value = stats.shapiro(df_cost['cost'])

# round funtion to round off 3 digits
p_value = round(p_value,3)

# print the p-value
print('p-value:', p_value)

p-value: 0.911


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the cost is normally distributed. Thus the assumption of normality is satisfied.

Let us check the equality of variances.

In [44]:
# perform Levene's test for the equality of variances 
# levene() returns a tuple having the values of test statistics and the corresponding p-value
# pass the number of defective belts from each group
stat, p_value = stats.levene(df_cost[df_cost['medium'] == 'Snapmart']['cost'], 
                             df_cost[df_cost['medium'] == 'Picbook']['cost'],
                             df_cost[df_cost['medium'] == 'Funsta']['cost'],
                             df_cost[df_cost['medium'] == 'Klicker']['cost'])
# round funtion to round off 3 digits
p_value = round(p_value,3)

# print the p-value 
print('P-Value:', p_value)

P-Value: 0.723


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the population variances are equal for all the samples.

Let us perform one-way ANOVA.

In [45]:
# obtain number of unique advertising mediums
t = df_cost['medium'].nunique()
print('t:', t)

# consider the total number of observations 
N = df_cost['medium'].value_counts().sum()
print('N:', N)

t: 4
N: 12


For ⍺ = 0.01 and degrees of freedom (= t-1, N-t) = (3, 8), calculate the critical value.

In [46]:
# calculate the F-value for 99% of confidence level
# use 'stats.f.isf()' to find the F-value corresponding to the upper tail probability 'q'
# pass the value of 'alpha' to the parameter 'q', here alpha = 0.01
# pass the degrees of freedom (= 3) to the parameter 'dfn' 
# pass the degrees of freedom (= 8) to the parameter 'dfd' 
# use 'round()' to round-off the value to 4 digits
f = np.abs(round(stats.f.isf(q = 0.01, dfn = 3, dfd = 8), 4))

print('Critical value for F-test:', f)

Critical value for F-test: 7.591


In [47]:
# perform one-way ANOVA
# pass the given data
f_test, p_val = stats.f_oneway(df_cost[df_cost['medium'] == 'Snapmart']['cost'], 
                               df_cost[df_cost['medium'] == 'Picbook']['cost'],
                               df_cost[df_cost['medium'] == 'Funsta']['cost'],
                               df_cost[df_cost['medium'] == 'Klicker']['cost'])
# round funtion to round off 3 digits
f_test = round(f_test,3)

# print the test statistic
print('Test statistic:', f_test)

Test statistic: 14.88


The above output shows that the test statistic is greater than 7.591. Thus we reject the null hypothesis and conclude that the average cost for at least one advertising medium is different.

<table align="left">
    <tr>
        <td width="6%">
            <img src="question_icon.png">
        </td>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b>9. If null hypothesis is rejected,  find which medium's advertisting cost is significantly different.</b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [48]:
# perform tukey's range test to compare the mean daily rate for pair of business travels
# pass the daily rate to the parameter, 'data'
# pass the type of business travel to the parameter, 'groups'
comp = mc.MultiComparison(data = df_cost['cost'], groups = df_cost['medium'])

# tukey's range test
# pass the alpha value as 0.01 to get the 99% confidence interval 
post_hoc = comp.tukeyhsd(alpha = 0.01)

# print the summary table
post_hoc.summary()

<class 'statsmodels.iolib.table.SimpleTable'>

The `reject=False` for pair (Funsta, Klicker) denotes that we fail to reject the null hypothesis; and conclude that the Futsta and Kliker average advertising cost is not significantly differ.

The `reject=True` for the pairs (Picbook, Snapmart) denotes that we reject the null hypothesis; and conclude that the average addevrtising cost  is significantly differ.